### PR TITLE
{ai}[foss/2022a] PyTorch v1.13.1 w/ Python 3.10.4 w/ CUDA 11.7.0

### PR DESCRIPTION
**Pull request by branfosj**
(created using `eb --new-pr`)

* [x] #17295
* [x] #17155

### CODE REVIEW

**Comment by satishskamath** [comment]
Hi @branfosj. Are there still issues pending for within this PR?

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
taurusi8006 - Linux CentOS Linux 7.9.2009, x86_64, AMD EPYC 7352 24-Core Processor (zen2), 8 x NVIDIA NVIDIA A100-SXM4-40GB, 470.57.02, Python 2.7.5
See https://gist.github.com/d562b0460290df6c5c3cde89694c1311 for a full test report.

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
taurusml26 - Linux RHEL 7.6, POWER, 8335-GTX (power9le), 6 x NVIDIA Tesla V100-SXM2-32GB, 440.64.00, Python 2.7.5
See https://gist.github.com/24a30bba9da16fb44941b172986475a0 for a full test report.

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
taurusa11 - Linux CentOS Linux 7.7.1908, x86_64, Intel(R) Xeon(R) CPU E5-2603 v4 @ 1.70GHz (broadwell), Python 2.7.5
See https://gist.github.com/bb99683f64fd9982ae70752654544d3d for a full test report.

**Comment by smoors** [comment]
Test report by @smoors
**FAILED**
Build succeeded for 2 out of 3 (1 easyconfigs in total)
node406.hydra.os - Linux CentOS Linux 7.9.2009, x86_64, AMD EPYC 7282 16-Core Processor, 1 x NVIDIA NVIDIA A100-PCIE-40GB, 515.48.07, Python 3.6.8
See https://gist.github.com/smoors/cefd98a7e9a2da2b2683ffecc386d663 for a full test report.

**Comment by smoors** [comment]
as discussed in the last conf call, to avoid this PR becoming stale and as the number of failed tests is limited, we decided to merge this as is and create an issue for the failing tests to follow-up.

@branfosj if you agree, can you add `max_failed_tests = 10` and remove the `draft` label? 
i'll then merge this and create the issue for the failing tests.

**Comment by boegel** [comment]
@smoors Issue already created, see https://github.com/easybuilders/easybuild-easyconfigs/issues/17712 (I'm focusing on getting https://github.com/easybuilders/easybuild-easyconfigs/pull/17155 merged first)

**Comment by boegel** [comment]
For `test_ops_gradients`, we probably just need to add the `PyTorch-1.13.1_skip-failing-grad-test.patch` as was done in https://github.com/easybuilders/easybuild-easyconfigs/pull/17155/commits/6124d4caf74f2d2a9f1a936586528d900e5d7324 in #17155

For the `test_jit*` failing tests, we should include those in `excluded_tests`, which is a bit more strict than just allowing 10 random tests to fail.

**Comment by branfosj** [comment]
Yes, we should get #17155 merged first and then make sure all the patches are synced to here.

**Comment by boegel** [comment]
Failing tests on POWER9:
```
test_ao_sparsity failed!
test_optim failed!
test_quantization failed!
distributed/rpc/test_tensorpipe_agent failed!
test_cpp_extensions_aot_ninja failed!
test_cpp_extensions_aot_no_ninja failed!
test_cpp_extensions_open_device_registration failed!
test_cuda failed!
test_ops failed!
```

Let's not block this PR over that, those can be dealt with in a follow-up PR.

**Comment by smoors** [comment]
> For the `test_jit*` failing tests, we should include those in `excluded_tests`, which is a bit more strict than just allowing 10 random tests to fail.

true, but on the other hand i prefer to run a test and ignore the failure than to skip the test altogether, especially if the failure is specific to another architecture than the one i am building on.

we could add another parameter `ignored_tests`, but that may be overkill..

**Comment by VRehnberg** [comment]
Test report by @VRehnberg
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
alvis2-12 - Linux Rocky Linux 8.6, x86_64, Intel(R) Xeon(R) Gold 6226R CPU @ 2.90GHz, 8 x NVIDIA Tesla T4, 520.61.05, Python 3.6.8
See https://gist.github.com/VRehnberg/eb6da3f62c2c703ca377f93735d71cc1 for a full test report.

**Comment by VRehnberg** [comment]
I'm seeing three failed tests:
```
Failed tests (suites/files):
* distributed/_shard/sharded_tensor/ops/test_linear
* distributed/rpc/test_tensorpipe_agent
* test_jit_legacy
distributed/_shard/sharded_tensor/ops/test_linear (3 total tests, errors=1)
```

For distributed/_shard/sharded_tensor/ops/test_linear there is one bfloat16 tensor comparison where 385.5 is compared to 385 in a few different places for 3 out of four GPUs (this is on the 4xA40 node).

This is not much for bfloat16 so I'd say either skip test or remove absolute tolerance and increase relative tolerance to at least $(2^0 + 2^-7)/2^0$ which should be about as much accuracy to expect with a 7-bit mantissa. However, I'm also confused because unless I'm miscounting the closest bfloat16 numbers should be 384 and 386 (i.e. 385 and 385.5 shouldn't be expressible) so there might be something I'm missing.

I'll check the other tests next.

**Comment by branfosj** [comment]
Test report by @branfosj
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
bear-pg0203u29a.bear.cluster - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), 1 x NVIDIA NVIDIA A100-SXM4-80GB, 520.61.05, Python 3.6.8
See https://gist.github.com/branfosj/e3f2805e2bf1d4e69c34a49bb6d3a671 for a full test report.

**Comment by boegel** [comment]
@branfosj Should be synced with `develop` now that #17155 is merged

**Comment by branfosj** [comment]
## `test_jit_legacy`, `test_jit_profiling`, and `test_jit`

Same failed test in the three testsuites

```
======================================================================
FAIL: test_freeze_conv_relu_fusion (jit.test_freezing.TestFrozenOptimizations)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/dev/shm/branfosj/build-up-EL8/PyTorch/1.13.1/foss-2022a-CUDA-11.7.0/pytorch-v1.13.1/test/jit/test_freezing.py", line 2258, in test_freeze_conv_relu_fusion
    self.assertEqual(mod_eager(inp), frozen_mod(inp))
  File "/dev/shm/branfosj/tmp-up-EL8/eb-8fe5hmfe/tmpuilxmqz9/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2470, in assertEqual
    assert_equal(
  File "/dev/shm/branfosj/tmp-up-EL8/eb-8fe5hmfe/tmpuilxmqz9/lib/python3.10/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 10 / 30 (33.3%)
Greatest absolute difference: 3.057718276977539e-05 at index (2, 3, 0, 0, 0) (up to 1e-05 allowed)
Greatest relative difference: 8.758584417742737e-05 at index (0, 3, 0, 0, 0) (up to 1.3e-06 allowed)

----------------------------------------------------------------------
```

## `test_optim`

```
======================================================================
FAIL: test_rprop (__main__.TestOptim)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/dev/shm/branfosj/tmp-up-EL8/eb-8fe5hmfe/tmpuilxmqz9/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 1054, in wrapper
    fn(*args, **kwargs)
  File "/dev/shm/branfosj/build-up-EL8/PyTorch/1.13.1/foss-2022a-CUDA-11.7.0/pytorch-v1.13.1/test/test_optim.py", line 1016, in test_rprop
    self._test_basic_cases(
  File "/dev/shm/branfosj/build-up-EL8/PyTorch/1.13.1/foss-2022a-CUDA-11.7.0/pytorch-v1.13.1/test/test_optim.py", line 283, in _test_basic_cases
    self._test_state_dict(
  File "/dev/shm/branfosj/build-up-EL8/PyTorch/1.13.1/foss-2022a-CUDA-11.7.0/pytorch-v1.13.1/test/test_optim.py", line 258, in _test_state_dict
    self.assertEqual(bias, bias_cuda)
  File "/dev/shm/branfosj/tmp-up-EL8/eb-8fe5hmfe/tmpuilxmqz9/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2470, in assertEqual
    assert_equal(
  File "/dev/shm/branfosj/tmp-up-EL8/eb-8fe5hmfe/tmpuilxmqz9/lib/python3.10/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 1 / 10 (10.0%)
Greatest absolute difference: 0.00010061264038085938 at index (0,) (up to 1e-05 allowed)
Greatest relative difference: 6.106863159088995e-05 at index (0,) (up to 1.3e-06 allowed)

----------------------------------------------------------------------
```

**Comment by branfosj** [comment]
Test report by @branfosj
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
bear-pg0203u29a.bear.cluster - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), 1 x NVIDIA NVIDIA A100-SXM4-80GB, 520.61.05, Python 3.6.8
See https://gist.github.com/branfosj/8a3001aa6da37bfc32277f9cf3eb9ecb for a full test report.

File: easybuild/easyconfigs/p/PyTorch/PyTorch-1.13.1-foss-2022a-CUDA-11.7.0.eb [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    {'PyTorch-1.13.1_remove-flaky-test-in-testnn.patch':
+     'be83ff61fe2dedab6d49c232936d5622df81ab49154264490021c6c828e53315'},
+    {'PyTorch-1.13.1_skip-ao-sparsity-test-without-fbgemm.patch':
+     '92cd48ef6d01aa7e07ccce1dcaf40bc3fb0f220c4aa4fea15f3e05fb42e37909'},
+    {'PyTorch-1.13.1_increase-tolerance-test_ops.patch':
+     'd53e98bf0da7788b68042dcc31bc5708dae962fde3f110cc827eb807a5d08e49'},
+    {'PyTorch-1.13.1_install-vsx-vec-headers.patch':
+     '7b678f54bb947afd4767f5877ac424b4b94ce5db609ea20f5a869ccf4027035f'},
+    {'PyTorch-1.13.1_skip-failing-grad-test.patch': '6681200f9509893cb9231b5c93ac9bc5e6d9d9ae4febefca52e7cbc843ba8f51'},
+    {'PyTorch-1.13.1_skip-test_freeze_conv_relu_fusion.patch',
```
Comment (branfosj) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-1.13.1-foss-2022a-CUDA-11.7.0.eb:69]:
```suggestion
    {'PyTorch-1.13.1_skip-test_freeze_conv_relu_fusion.patch':
```

**Comment by branfosj** [comment]
Test report by @branfosj
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
bear-pg0203u29a.bear.cluster - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), 1 x NVIDIA NVIDIA A100-SXM4-80GB, 520.61.05, Python 3.6.8
See https://gist.github.com/branfosj/a91c14fec2ac001a1831a8229e17a14a for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
bear-pg0203u31a.bear.cluster - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), 4 x NVIDIA NVIDIA A100-SXM4-40GB, 520.61.05, Python 3.6.8
See https://gist.github.com/branfosj/11e97ea919a6ce9e6b191b8a1c5870b6 for a full test report.

**Comment by branfosj** [comment]
closing for #18305

**Pull request closed without merging**